### DISCUSSION TITLE
Plotting Runs, HellaSwag Eval, and Text Generation

### DISCUSSION THREAD

**Discussion by chrisjmccormick** [KellerJordan/modded-nanogpt#156] [Show and tell]
I put together another Colab Notebook, [Modded-NanoGPT - v2025-10-27.ipynb](https://colab.research.google.com/drive/1kv6G54avvUSTHi8zYXd_Lz1Gtu4VstDB#scrollTo=2908ccbb), which matches the [2025-10-27](https://github.com/KellerJordan/modded-nanogpt/blob/master/records/track_1_short/2025-10-27_FixMuonLR/14afd380-d3d9-48d7-ad23-4c13cb96754b.txt) record code and has a number of added features that may be useful for experimentation.

The current notebook output is from a 40GB A100 on Colab. Training time was just 45 minutes, and loss is `3.27535`. 

### Log Files and Plots

One of the last sections, "Results and Analysis", contains code for parsing the validation loss and training times from one or more `.txt` log files, and then generates some useful plots. For example:

<img src='https://lh3.googleusercontent.com/d/1FghiVeSkyH-RY8PNUBckNk3dGGHSLEVL' alt='Example plot of validation loss' width='900' />

(Note--I currently have it configured for running validation every 100 steps. Official script default is 250.)

### Top-Level Configuration

I created a `Configuration` dataclass at the top which captures / exposes more of the settings (such as learning rates and model dimensions). This can make it a little easier to review these settings and adjust them between experiments.

### Weights and Biases Integration

I added W&B logging to the training loop. You can disable it in the top level config, and / or it shouldn't be too hard to rip out if you don't like it. :)

For example: https://wandb.ai/chrismccormick/Modded-NanoGPT

### HellaSwag Evaluation

At the end of the Notebook is a section for running the HellaSwag benchmark. (This required adding an "inference" flag to GPT.forward).

Current result:
`10042 acc: 0.2813 acc_norm: 2927/10042=0.2915`

### Pre-Calculated Schedules

After the configuration section, I pre-calculate the schedules for momentum, learning rates, and window sizes.

This:
* Allows for plotting the schedules prior to running the code
* Makes it easier to resume from a checkpoint (not implemented)
* Makes it easier to hack on / customize the schedules

<img src='https://lh3.googleusercontent.com/d/1jGHytqAls4NqsBkbicMDkskn19i8q7Co' alt='Example plot of a modified speed run training schedule' width='900' />

### Single GPU Support

I run this Notebook on Colab A100s and GH200s at Lambda Labs.

It's designed to run in both those environments without any additional setup--just upload and hit run!

(Ok, you'll still need to provide API keys, and update your wandb username, but no docker or virtual environment required!)

### Text Generation from a Prompt

The final section of the Notebook demonstrates how to reload the model and give it a prompt to generate from. It was tricky to get the details of this right, but it seems to be working well.

```python
generated_text = generate_padded(
    prompt=" The quick brown fox",
    max_new_tokens=30,
    temperature=0.7
)
```

```
 The quick brown fox has been a popular addition to the kitchen table - and a great option for those who are more adventurous.
This multi-purpose brown fox is also
```

```python
generated_text = generate_padded(
    prompt=" In a surprising turn of events,",
    max_new_tokens=30,
    temperature=0.7
)
```

```
 In a surprising turn of events, the NFL team's best player in the league, Josh McDaniels, threw a tantrum to his teammates just days before the game.
Mc
```
